spec/editors_draft.html

```diff
@@ -3298,6 +3298,231 @@ <h2>Evaluation Semantics</h2>
+      <pre id="load-renaming-blank-example-consistent-ids"
+         title="Producing consistent blank node identifiers shared between RDF and Composite Values"
+         data-transform="updateExample"
+         data-content-type="text/turtle"
+         class="nohighlight example">
+      <!--
+      _:b0 ex:p3 "[ _:b0 ]"^^cdt:List .
+      -->
+      </pre>
+      <pre id="load-renaming-blank-example-consistent-ids-result"
+         title="Producing consistent blank node identifiers shared between RDF and Composite Values: Result"
+         data-transform="updateExample"
+         data-content-type="text/n-triples"
+         class="nohighlight example">
+      <!--
+      _:b004 <http://example.org/p3> "[ _:b004 ]"^^<http://w3id.org/awslabs/neptune/SPARQL-CDTs/List> .
+      -->
+      </pre>
+
+      <p>Loading the Turtle in <a href="#load-renaming-blank-reuse"></a> with a triple with a <a>cdt:Map literal</a> containing two map entries, both having the same blank node value, MUST result in a dataset with a single triple whose <a>cdt:Map literal</a> object contains two map entries having the same blank node value, and whose identifier MUST NOT be "b0" (<a href="#load-renaming-blank-reuse-result"></a>):</p>
+      <pre id="load-renaming-blank-reuse"
+         title="Producing consistent blank node identifiers used multiple times in Composite Values"
+         data-transform="updateExample"
+         data-content-type="text/turtle"
+         class="nohighlight example">
+      <!--
+      ex:s ex:p3 "{ 'this': _:b0, 'again': _:b0 }"^^cdt:Map .
+      -->
+      </pre>
+      <pre id="load-renaming-blank-reuse-result"
+         title="Producing consistent blank node identifiers used multiple times in Composite Values: Result"
+         data-transform="updateExample"
+         data-content-type="text/n-triples"
+         class="nohighlight example">
+      <!--
+      <http://example.org/s> <http://example.org/p3> "{'this': _:b005,'again': _:b005}"^^<http://w3id.org/awslabs/neptune/SPARQL-CDTs/Map> .
+      -->
+      </pre>
+    </section>
+
+    <section>
+      <h2>Exporting Requirements</h2>
+      <p>
+        Conforming implementations MUST process <a>cdt:List literal</a>s and <a>cdt:Map literal</a>s during export, replacing the substrings containing the serialized form of any blank nodes with identifier <var>id</var> with a blank node label <var>b</var>, such that
+      </p>
+
+      <ul>
+        <li>within the scope of the document being serialized, all occurrences of <var>id</var> as a blank node identifier are replaced with the same <var>b</var> value (both inside and outside of composite values), and</li>
+        <li>within the scope of the document being serialized, two distinct <var>id</var> values are replaced with distinct <var>b</var> values.</li>
+      </ul>
```

Comment:
Also for these two bullet points, I somehow have the feeling that something is missing here. For instance, it seems to me that I would be conforming to these two rules if I was simply using _b_=_id_ for each of the blank node identifiers, but I guess that this is not your intention.

Additionally, specifically for export, I believe a distinction needs to be made between systems that store literals internally based on the lexical forms versus systems that store literals internally based on the values represented by the literals. The bullet points here seem to focus on the former type of systems. Is there anything to be said about the latter type of systems?

Reply:
I think you *would* be conforming if you used `b=id`. My intention with this section was to ensure that the serialization of a blank node that appears both inside and outside a composite value will end up with the same identifier. Since many RDF serializers produce simple bnode identifiers during serialization, we have to ensure that the blank nodes in the composite literals are treated the same.

Can you expand a bit on the second issue you ask about here? I think my two suggested bullet point rules would apply to both system types, but I'm not sure I'm fully understanding.

Reply:
Okay, let me try to elaborate more on my thinking regarding the second issue: The text and the two bullet points here talk about replacing "substrings" (which represent blank node identifiers). Such substrings may occur only in the lexical form of CDT literals, not in the value form. Consequently, replacing such substrings is something that is only relevant in a context in which a system deals with the lexical form of the literals. Now, systems that store literals internally based on the lexical form would naturally have to deal with these lexical forms when exporting the data. In contrast, systems that store literals internally based on the values represented by the literals would not have any lexical form to start with, when exporting. The question then is: how are the two bullet points (about replacing substrings in lexical forms) relevant for the latter kind of systems?

Reply:
I don't think the implementation approach is relevant here. Even for systems that store literals internally based on their values still have to treat them as RDF literals, which in the [abstract syntax](https://www.w3.org/TR/rdf11-concepts/#section-Graph-Literal) have "a lexical form, being a Unicode [UNICODE] string". That's the thing that a serializer is operating on conceptually, regardless of how it might actually be implemented.

Reply:
I realized that the export rules were missing a condition that would keep blank nodes appearing both inside and outside of composite values aligned with their serialized identifiers. Newly added rule:

* if the format of the document being serialized has a mechanism for explicitly identifying blank nodes outside of cdt:List and cdt:Map literals (e.g., _:b1 syntax in N-Triples and in Turtle, rdf:nodeID in RDF/XML), and the blank node B is contained in the data to be serialized both inside and outside of composite value literals, then the serializer for this document format MUST serialize <var>B</var> using the blank node identifier <var>b</var> (both inside and outside of composite values)

Reply:
While I generally agree with the new (third) bullet point that you have added here, this point feels somewhat detached from the context in which you are listing these bullet points. I mean, the context is that there is some substring "`_:`_id_" in the lexical form of some cdt:List or cdt:Map literal to be exported, but this bullet point does not talk at all about this substring / about _id_. In fact, within this bullet point you seem to assume systems that represent CDT literals internally based their value form (how would such a system otherwise know that it is the same blank node B that is both inside and outside of CDT literals?).

There is also another potential problem that just came to my mind when reading this new (third) bullet point: While the point focuses on cases in which the format of the document being serialized has a mechanism for explicitly identifying blank nodes outside of CDT literals, I wonder what a system should do for formats that do not have such a mechanism. Assume the dataset in a system contains the triple (_b_, :p, _l_) where _b_ is a blank node and _l_ is a cdt:List literal whose value (i.e., term list) contains _b_ as well. If the dataset with this triple needs to be serialized into a format that does not provide such an aforementioned mechanism, then there is no way to serialize the triple without loosing the connection between the blank node in the subject and the blank node inside the object literal, right?